crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -644,235 +722,248 @@ impl BlankLinesChecker {
-        stylist: &Stylist,
         diagnostics: &mut Vec<Diagnostic>,
     ) {
-        self.class_status.update(line);
-        self.fn_status.update(line);
-
-        // Don't expect blank lines before the first non comment line.
-        if self.is_not_first_logical_line {
-            if line.preceding_blank_lines == 0
+        if line.preceding_blank_lines == 0
                 // Only applies to methods.
                 && matches!(line.kind,  LogicalLineKind::Function | LogicalLineKind::Decorator)
                 // Allow groups of one-liners.
-                && !(matches!(self.follows, Follows::Def) && !matches!(line.last_token, TokenKind::Colon))
-                && matches!(self.class_status, Status::Inside(_))
+                && !(matches!(state.follows, Follows::Def) && !matches!(line.last_token, TokenKind::Colon))
+                && matches!(state.class_status, Status::Inside(_))
                 // The class/parent method's docstring can directly precede the def.
                 // Allow following a decorator (if there is an error it will be triggered on the first decorator).
-                && !matches!(self.follows, Follows::Docstring | Follows::Decorator)
+                && !matches!(state.follows, Follows::Docstring | Follows::Decorator)
                 // Do not trigger when the def follows an if/while/etc...
                 && prev_indent_length.is_some_and(|prev_indent_length| prev_indent_length >= line.indent_length)
-            {
-                // E301
-                let mut diagnostic =
-                    Diagnostic::new(BlankLineBetweenMethods, line.first_token_range);
-                diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
-                    stylist.line_ending().to_string(),
-                    locator.line_start(self.last_non_comment_line_end),
-                )));
+        {
+            // E301
+            let mut diagnostic = Diagnostic::new(BlankLineBetweenMethods, line.first_token_range);
+            diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                self.stylist.line_ending().to_string(),
+                self.locator.line_start(state.last_non_comment_line_end),
+            )));
+
+            diagnostics.push(diagnostic);
+        }
 
-                diagnostics.push(diagnostic);
+        let expected_blank_lines_before_definition = if line.indent_length == 0 {
+            // Mimic the isort rules for the number of blank lines before classes and functions
+            if state.follows.is_any_import() {
+                // Fallback to the default if the value is too large for an u32 or if it is negative.
+                // A negative value means that isort should determine the blank lines automatically
+                // Defaulting to 2 is correct because the variable is only used when testing the
+                // blank lines before a class or function definition.
+                u32::try_from(self.lines_after_imports).unwrap_or(BLANK_LINES_TOP_LEVEL)
```

Comment:
I'm not sure I understand this justification. I would have expected the default/automatic value from isort instead of `BLANK_LINES_TOP_LEVEL`.

Reply:
I expanded the comment explaining the `isort` logic. `isort` defaults to 2 if before a class or function definition but that wasn't clear from my original comment.